### ISSUE TITLE
React-Calendar : error when deleting recurrence events

### ISSUE THREAD

**Issue by mohammadamer** [pnp/sp-dev-fx-webparts#1773]
## Sample
> React-Calendar

## Authors
@joaojmendes
@derhallim 
@Abderahman88

## Expected or Desired Behavior
> the excepted behavior is to see recurrence events deleted after select the event and click delete.

## Observed Behavior
> Recurrence events are not deleted after click delete. error appeared at the top of the edit panel
`Error making HttpClient request in queryable [400] ::> {"odata.error":{"code":"-1, Microsoft.SharePoint.Client.InvalidClientQueryException","message":{"lang":"en-US","value":"String was not recognized as a valid DateTime."}}}`

## Steps to Reproduce

1.Create a recurrence event 
2.Go to the new created event
3. Click on 'delete'
`Error making HttpClient request in queryable [400] ::> {"odata.error":{"code":"-1, Microsoft.SharePoint.Client.InvalidClientQueryException","message":{"lang":"en-US","value":"String was not recognized as a valid DateTime."}}}`
5.You will notice that event is not deleted

# Environment Details (*Development & Target environment*)

- **OS**: [e.g. Windows 10 | 
- **Target Environment**: [ SharePoint Online]
- **Framework**: [ Node.js v10.24]
- **Browser(s)**: [ Edge]
- **Tooling**: [e.g. VS Code | SPFx v1.10.0 ]
- **Additional details**: *The more context you can provide, the easier it is (and therefore quicker) to help.*

**Comment by Abderahman88** [comment]
Hi @mohammadamer 

As I can remember, there are 2 ways to delete recurrent events:

**1. Delete all recurrent events**
a) Open a recurrent event
b) Click on `Edit Recurrence Series`
c) Click on `Delete`

This will delete all events. (Tested and works)

![RecEventDem](https://user-images.githubusercontent.com/36161889/111346840-e7546080-867e-11eb-8498-77ad9126af1f.png)

**2. Delete one event in the recurrent series**
a) Open a recurrent event
b) Click on `Delete`

Normally only that event will be deleted in the series and here is where the issue occurs. 
(I'm pretty sure this used to work 🙄)

![DelEventError](https://user-images.githubusercontent.com/36161889/111347468-88431b80-867f-11eb-83dc-a77033277e69.png)


After investigation, the change need to be made in this part of the code.
I think one of those parameters is not converted well to a datetime-string.

```
 public async addEvent(newEvent: IEventData, siteUrl: string, listId: string) {
    let results = null;
    try {
      const web = new Web(siteUrl);

      results = await web.lists.getById(listId).items.add({
        Title: newEvent.title,
        Description: newEvent.Description,
        Geolocation: newEvent.geolocation,
        ParticipantsPickerId: { results: newEvent.attendes },
        EventDate: await this.getUtcTime(newEvent.EventDate),
        EndDate: await this.getUtcTime(newEvent.EndDate),
        Location: newEvent.location,
        fAllDayEvent: newEvent.fAllDayEvent,
        fRecurrence: newEvent.fRecurrence,
        Category: newEvent.Category,
        EventType: newEvent.EventType,
        UID: newEvent.UID,
        RecurrenceData: newEvent.RecurrenceData ? await this.deCodeHtmlEntities(newEvent.RecurrenceData) : "",
        MasterSeriesItemID: newEvent.MasterSeriesItemID,
        RecurrenceID: newEvent.RecurrenceID ? await this.getUtcTime(newEvent.RecurrenceID) : undefined,
      });
    }
    catch (error) {
      return Promise.reject(error);
    }
    return results;
  }
```

**Comment by mohammadamer** [comment]
Hi @Abderahman88,

Can you please explain why when delete only one occurrence, A new event should be created.

I do not see any functionality about deletion in that case 
deleteEvent(...)
{
.....
else {
            **// delete a single recurrence Exception.  add new entry with eventtype 3**

            event.RecurrenceID = event.EventDate.toString();
            event.MasterSeriesItemID = event.ID.toString();
            event.fRecurrence = true;
            event.EventType = '3';
            await this.addEvent(event, siteUrl, listId);
          }
.....
}

Regardless that the confirmation message about delete all occurrence is shown up in both cases (one occurrence & all occurrence ). I'll look into it.

**Comment by Abderahman88** [comment]
@mohammadamer 

I'm not exactly sure why a new event should be created, but there will be some reason for it.
Maybe it has to do with the way events are handled by SharePoint.

For the confirmation message, you're right.
If you delete one occurence, the popup should not appear or show another message for example 'Only this event will be deleted ...'

I also found out when creating recurrent events, the dates are not always correct.
In a nutshell, there's still a lot of room of improvement.

**Comment by mohammadamer** [comment]
Hi @Abderahman88

I see that what you are explaining is implemented in the the code but it's implemented when the **event type is equal 4**  thanks for your investigations. good to know that.
But in case of deletion of one occurrence and **event type is equal 1**
> a new event is created without (Deleted) as prefix
> set RecurrenceID to event date
>set MasterSeriesItemID to item id
> set fRecurrence to true
>set EventType to 3
_This part of the code is not clear for me._

I did some debugging and found out that the issue of date is in the conversion of date string in this line of code
```
 RecurrenceID: newEvent.RecurrenceID
          ? await this.getUtcTime(newEvent.RecurrenceID)
          : undefined,
```
I avoided this date time exception by setting RecurrenceID to null to avoid date time string conversion and to be able to investigate the issue but still working on it.

**Comment by Abderahman88** [comment]
Hi @mohammadamer 

By creating a recurrent event, the event type is set to **1** (see example)

![RecEvent](https://user-images.githubusercontent.com/36161889/111387792-98262400-86ae-11eb-92ca-e2debf92728a.png)

When deleting one of the recurring events, it checks for the value of `recurrenceSeriesEdited`.
Standard this is set to `false` but clicking on 'Edit Recurrence Series' -button this is set to `true`. (Which essencially means, go to the master event)
When `recurrenceSeriesEdited = true` then the master-event will be deleted, so all recurrent events are gone.
When` recurrenceSeriesEdited = false` only one single recurrent event needs to be deleted. This entails the creation of a new event with `event type = 3` (see screenshot in previous post)
This is done by `this.addEvent(event, siteUrl, listId);`

At the momet the Deleted-prefix is not added to the Title. We could change this, to stay in sync with standard Sharepoint.
```
case '1': // recurrence Event
          // if  delete is a main recrrence delete all recurrences and main recurrence
          if (recurrenceSeriesEdited) {
            // delete execptions if exists before delete recurrence event
            await this.deleteRecurrenceExceptions(event, siteUrl, listId);
            await web.lists.getById(listId).items.getById(event.Id).delete();
          } else {
            // delete a single recurrence Exception.  add new entry with eventtype 3

            event.RecurrenceID = event.EventDate.toString();
            event.MasterSeriesItemID = event.ID.toString();
            event.fRecurrence = true;
            event.EventType = '3';
            await this.addEvent(event, siteUrl, listId);
          }
```
For the second point:

You are correct. The problem lies here :

```
 RecurrenceID: newEvent.RecurrenceID
          ? await this.getUtcTime(newEvent.RecurrenceID)
          : undefined,
```

`await this.getUtcTime(newEvent.RecurrenceID)` returns a string.
This needs to be converted to a dateTime to solve it.

![recDateTime](https://user-images.githubusercontent.com/36161889/111390149-89da0700-86b2-11eb-9b0b-01521f730a40.png)

**Comment by mohammadamer** [comment]
> @mohammadamer
> 
> I'm not exactly sure why a new event should be created, but there will be some reason for it.
> Maybe it has to do with the way events are handled by SharePoint.
> 
> For the confirmation message, you're right.
> If you delete one occurence, the popup should not appear or show another message for example 'Only this event will be deleted ...'
> 
> I also found out when creating recurrent events, the dates are not always correct.
> In a nutshell, there's still a lot of room of improvement.
> 
> **UPDATE**
> Deleting one event from a recurrent series will result in the creation of a new event. This is standard SharePoint behavior.
> The new event gets the original title with 'Deleted' as prefix and the event type is set to 3.
> ![deletedEvent](https://user-images.githubusercontent.com/36161889/111373657-79b72d00-869c-11eb-940a-d3929db931f8.png)

**Comment by mohammadamer** [comment]
> @mohammadamer
> 
> I'm not exactly sure why a new event should be created, but there will be some reason for it.
> Maybe it has to do with the way events are handled by SharePoint.
> 
> For the confirmation message, you're right.
> If you delete one occurence, the popup should not appear or show another message for example 'Only this event will be deleted ...'
> 
> I also found out when creating recurrent events, the dates are not always correct.
> In a nutshell, there's still a lot of room of improvement.
> 
> **UPDATE**
> Deleting one event from a recurrent series will result in the creation of a new event. This is standard SharePoint behavior.
> The new event gets the original title with 'Deleted' as prefix and the event type is set to 3.
> ![deletedEvent](https://user-images.githubusercontent.com/36161889/111373657-79b72d00-869c-11eb-940a-d3929db931f8.png)

Hey @Abderahman88 :)
Regarding the update the you provide above.
Could you please elaborate how did you do (Deleting one event from a recurrent series in the SharePoint Standard functionality).

**Comment by mohammadamer** [comment]
Hi @Abderahman88

I have done some changes:
1. I've changed the RecurrentID to Date type
2. I've done some fixes for RecurrentID in some places since it became a Date Type
3. I've applied the same SharePoint Standard functionality for deleting one occurrence of recurrent event by adding prefix "Deleted"
It works fine :)
I'll add a pull request with these changes.

**Comment by Abderahman88** [comment]
@mohammadamer 

Great job 👍
